pkg/cloud/ratelimit.go

```diff
@@ -112,3 +114,148 @@ func (m *MinimumRateLimiter) Accept(ctx context.Context, key *RateLimitKey) erro
 func (m *MinimumRateLimiter) Observe(ctx context.Context, err error, key *RateLimitKey) {
 	m.RateLimiter.Observe(ctx, err, key)
 }
+
+// TickerRateLimiter uses time.Ticker to spread Accepts over time.
+//
+// Concurrent calls to Accept will block on the same channel. It is not
+// guaranteed what caller will be unblocked first.
+type TickerRateLimiter struct {
+	ticker *time.Ticker
+}
+
+// NewTickerRateLimiter creates a new TickerRateLimiter which will space Accept
+// calls at least interval/limit time apart.
+//
+// For example, limit=4 interval=time.Minute will unblock a single Accept call
+// every 15 seconds.
+func NewTickerRateLimiter(limit int, interval time.Duration) *TickerRateLimiter {
+	return &TickerRateLimiter{
+		ticker: time.NewTicker(interval / time.Duration(limit)),
+	}
+}
+
+// Accept will block until a time, specified when creating TickerRateLimiter,
+// passes since the last call to Accept.
+func (t *TickerRateLimiter) Accept(ctx context.Context, rlk *RateLimitKey) error {
+	select {
+	case <-t.ticker.C:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Observe does nothing.
+func (*TickerRateLimiter) Observe(context.Context, error, *RateLimitKey) {
+}
+
+// Make sure that TickerRateLimiter implements RateLimiter.
+var _ RateLimiter = new(TickerRateLimiter)
+
+// CompositeRateLimiter combines rate limiters based on RateLimitKey.
+type CompositeRateLimiter struct {
+	// map[project id]map[resource name]map[operation name]RateLimiter
+	rateLimiters map[string]map[string]map[string]RateLimiter
+	// defaultRL is used when no matching RateLimiter was found.
+	defaultRL RateLimiter
+}
+
+// NewCompositeRateLimiter creates a new CompositeRateLimiter that will use
+// provided default rate limiter if no better match is found.
+//
+// # Example
+//
+//	defaultRL := /* default rate limiter */
+//	bsGetListRL := /* backend service rate limiter for get and list operation in project-1 */
+//	projectRL := /* rate limiter for project-1 */
+//	bsOtherProjectRL := /* rate limiter for backend service in other projects */
+//
+//	rl := NewCompositeRateLimiter(defaultRL)
+//	rl.Register("project-1", "", "", projectRL)
+//	rl.Register("project-1", "BackendServices", "Get", bsGetListRL)
+//	rl.Register("project-1", "BackendServices", "List", bsGetListRL)
+//	rl.Register("", "BackendServices", "", bsOtherProjectRL)
```

Comment:
From business point of view is there a need to share one rate limiter across project? I can see in the example that you are not sharing but why you accept empty project name? This implies that we can share RL across projects and IMO it's not correct.

Reply:
> From business point of view is there a need to share one rate limiter across project?

Yes, quotas are very often defined for a set of resources and/or operations.

> why you accept empty project name?

Providing an empty project name is a way to set a "default" if there is no more specific rate limiter provided. This is explained in the other docstring comment.

Reply:
I check public docs and from what I understand all quotas are defined per project https://screenshot.googleplex.com/7dN4SBk2oamcguC. So you should not share RateLimiter across projects. You should set default RL in context of projects. That means that every project should have independent instance of default RL. Projects passed to client are well defined so this code doesn't have to be super generic and handle all cases. Using this RL should be correlated with ProjectRouter because if you want to support multiple project you should support projects that you will use. (and in fact in most cases we are using Single project router or project router when we have network defined in different project)